refactor: 예외 핸들러 구조화 및, 누락된 기본 예외 핸들러 추가

cookie-meringue opened the pull request:
## As-Is
<!-- 문제 상황 정의 -->
예외 핸들링 로직이 하나의 클래스(`GlobalControllerAdvice`) 내에 전부 존재했음.
이로 인해 가독성이 저해되고, SRP 를 준수하지 못하고 있었음.


## To-Be
<!-- 변경 사항 -->
예상 가능한 예외와 예상 불가능한 예외를 추상화.
Additional Description 참조


## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?

## Test Screenshot


## (Optional) Additional Description

하나의 `@RestControllerAdvice` 클래스 아래에 있던 `@ExceptionHandler`들을 각각의 `@RestControllerAdvice - @ExceptionHandler` 쌍으로 분리.

**기존 구조**

```java
@RestControllerAdvice
class GlobalExceptionHandler {
|
|--- @ExceptionHandler(BusinessException.class)
|
|--- @ExceptionHandler(HandlerMethodValidationExcpetion.class)
|
|--- //...
|
|--- @ExceptionHandler(Exception.class)
}
```

**변경 후 구조**

```java
@RestControllerAdvice
class BusinessExceptionHandler {
|
|--- @ExceptionHandler(BusinessException.class)
}

@RestControllerAdvice
class HandlerMethodValidationExcpetionHandler {
|
|--- @ExceptionHandler(HandlerMethodValidationExcpetion.class)
}

@RestControllerAdvice
class UnexpectedExcpetionHandler {
|
|--- @ExceptionHandler(Exception.class)
}
```

### 장점

시그니처의 제약이 없음.

```java
@RestControllerAdvice
public class AExceptionHandler {

		// 매개변수로 아무것도 받지 않을 수 있음.
    @ExceptionHandler(AException.class)
    public ProblemDetail handle() {
        return ErrorCode.A.toProblemDetail();
    }
}

@RestControllerAdvice
public class BExceptionHandler {

		// 매개변수로 Exception을 받을 수 있음.
    @ExceptionHandler(BException.class)
    public ProblemDetail handle(BException exception) {
        ExceptionLogger.logExpectedException(exception);
        return exception.getErrorCode().toProblemDetail();
    }
}

@RestControllerAdvice
public class CExceptionHandler {

		// 매개변수로 Exception, HttpServletRequest를 받을 수 있음.
    @ExceptionHandler(CException.class)
    public ProblemDetail handle(CException exception, HttpServletRequest req) {
        ExceptionLogger.logExpectedException(exception);
        // req를 다루는 로직.
        return exception.getErrorCode().toProblemDetail();
    }
}
```

상속과 구현의 관계에서 벗어났기 때문에, 예외 핸들러 시그니처가 다를 수 있음.

### 한계

사실상 클래스 분리만 한 것이므로, 기존의 문제점을 전부 해결할 수 없음.

**해결할 수 없는 문제 1**

> `@ExceptionHandler` 내부의 예외 클래스와 메서드 파라메터의 클래스 타입을 일치시킬 수 없음.
> 
> 
> ```java
> @ExceptionHandler(IllegalArgumentException.class)
> public ProblemDetail handle(NullPointerException exception) {
> 	// handle
> }
> ```
> 
> 위 코드 동작 시, 어노테이션과 파라메터의 예외 타입이 맞지 않아 런타임에 예외가 발생함.(예외 핸들러에서 예외 발생)
> 
> AOP 등을 사용해 예외 핸들러에서 발생한 예외를 핸들링할 수는 있지만, 이런 상황이 발생하는 것은 운영 환경이 아닌 테스트 과정에서 전부 끝나야 한다고 생각함.
> 
> 어노테이션과 파라메터의 예외 타입이 동일하도록 강제할 수단이 존재하지 않음.
> 

**해결할 수 없는 문제 2**

> 한 예외에 대해 여러 `@ExceptionHandler`가 존재할 수 있음.
> 
> 
> Spring MVC 는 패키지 레벨에 따른 예외 핸들링을 구현할 수 있게 하기 위해 이를 허용함.
> 
> 그러나, 우리 프로젝트에서는 중앙 집중화된 방식으로 예외를 처리하도록 설계했음.
> 
> → 패키지 레벨에 따른 예외 핸들링이 불필요!
> 
> 이러한 상황에서, 한 예외에 대해 여러 `@ExceptionHandler`가 존재할 수 있다는 점은 오히려 방해 요소가 됨.
> 
> 예외를 처리하는 패키지가 한 군데이기 때문.
> 
> 한 예외에 대해 여러 `@ExceptionHandler`를 만들면 예외 처리 동작을 파악하기 매우 힘듦.
> 

### 한계 해결

스프링 애플리케이션은 세 가지 시점이 있음.

1. 컴파일 시점
2. `ApplicationContext` 초기화 시점 (런타임 시점이나, 실질적 런타임 이전)
3. 실질적 런타임 시점

컴파일 시점에 검증을 수행할 수 없다면, `ApplicationContext` 초기화 시점에 검증하도록 구현.

실질적 런타임에 예외가 발생하는 것 보다는, `ApplicationContext` 초기화 실패로 애플리케이션이 구동되지 않도록 방지하는 구현을 선택.

Spring 의 `BeanPostProcessor`를 통해 빈 초기화 시 검증이 수행되도록 구현.

`@RestControllerAdvice`가 붙은 클래스의 메서드에 대해 다음과 같은 검증 프로세스 수행.

> 1. `@ExceptionHandler` 의 예외 클래스와 메서드의 파라메터에 있는 예외 클래스는 반드시 같아야 한다.
> 
> - 메서드가 예외 클래스를 파라메터로 받지 않는 것은 허용한다.
> - 만약, 메서드가 예외 클래스를 파라메터로 받는다면, 반드시 `@ExceptionHandler` 의 예외 클래스만을 받아야 한다.
> 
> 2. 애플리케이션 전역에서 특정 예외에 대한 `@ExceptionHandler` 는 단 하나만 존재해야 한다.
> 
> - `IllegalArgumentException`을 다루는 `@ExceptionHandler` 는 애플리케이션 전역에서 단 하나만 존재해야 한다.
> - 상속은 고려하지 않는다. (`IllegalArgumentException`을 다루는 핸들러와 `Exception`을 다루는 핸들러는 동시에 존재할 수 있다)

`BeanPostProcessor`를 통한 `ApplicationContext` 초기화 시점 검증을 통해 문제를 해결.

# 새로운 예외 핸들러를 추가하는 방법

`routie.global.exception.infrastructure.expected` 아래에 예외 클래스를 추가.

예외 클래스에는 반드시 `@RestControllerAdvice` 어노테이션을 사용.

혼동을 방지하기 위해 하나의 예외 클래스에는 하나의 `@ExceptionHandler`만 존재하도록 구현.

`IllegalArgumentException` 핸들러 추가 예시

```java
/**
* 설명
*/
@RestControllerAdvice
public class IllegalArgumentExceptionHandler {

    @ExceptionHandler(IllegalArgumentException.class)
    public ProblemDetail handle(final IllegalArgumentException exception) {
        ExceptionLogger.logExpectedException(exception);
				return ErrorCode.ILLEGAL_ARGUMENT.toProblemDetail();
    }
}
```

예외 클래스에 반드시 javadoc 을 통해 어떤 예외를 핸들링하는지, 예시 상황은 무엇이 있는지 작성해야 함.

# 기대 효과

## 1. 가독성

`@ExceptionHandler` 를 디렉터리 구조로 관리할 수 있어 가독성과 명시성이 매우 높아짐.

<img width="456" height="433" alt="image" src="https://github.com/user-attachments/assets/5a6e534a-369f-4762-89e2-b02103b5a126" />

각 예외 타입별로 패키지 분리 가능.

하나의 클래스에서 특정 예외 핸들러를 찾는 것 보다, 패키지와 클래스명을 통해 찾을 수 있어 더욱 편리함.

## 2. 런타임 예외 방지

하나의 예외에 대한 `@ExceptionHandler`가 겹침.

`@ExceptionHandler`의 예외 타입과 파라메터의 예외 타입이 다름.

`@ExceptionHandler` 메서드의 파라메터로 여러 `Exception`을 받음.

위 세 가지 경우, 스프링 애플리케이션이 실행될 때 아무런 문제가 없음.

그러나, 실질적으로 해당 `@ExceptionHandler`가 호출될 때 런타임 예외가 발생.

예외를 핸들링하는 핸들러에서 예외가 발생하는 것, 그리고 이를 런타임 전에 발견하지 못한 것은 치명적인 문제라고 판단.

`BeanPostProcessor` 를 통해 해당 상황들을 미연에 방지할 수 있음.

⇒ 런타임 예외 발생 전 사전에 판단할 수 있음(테스트가 실패할 것이기 때문)

cookie-meringue committed:
feat: 예외 정보를 담는 클래스 추가

cookie-meringue committed:
feat: 예외 Resolver 인터페이스 추가

cookie-meringue committed:
feat: 예상 가능한 예외 Resolver 추상 클래스 구현

cookie-meringue committed:
feat: 비즈니스 로직 수행 과정에서 발생하는 모든 예외를 처리하는 Resolver 구현

cookie-meringue committed:
feat: HandlerMethod 에 파라메터 바인딩 시, 검증에 실패한 경우 발생하는 예외를 처리하는 Resolver 구현

cookie-meringue committed:
feat: 요청 본문을 처리하는 과정에서 발생하는 예외를 처리하는 Resolver 구현

cookie-meringue committed:
feat: HandlerMethod 에 파라메터 바인딩 시, 타입이 일치하지 않는 경우 발생하는 예외를 처리하는 Resolver 구현

cookie-meringue committed:
feat: HandlerMethod 의 필수 요청 파라메터가 누락된 경우 발생하는 예외를 처리하는 Resolver 구현

cookie-meringue committed:
feat: 예상치 못한 예외를 처리하는 Resolver 구현

cookie-meringue committed:
feat: 예상 가능한, 예상하지 못한 예외를 처리하는 복잡한 로직을 추상화하기 위한 ExceptionResolver 파사드 구현

cookie-meringue committed:
feat: ExceptionResolver 파사드 의존성 주입 및 Primary Bean 등록 Config 작성

cookie-meringue committed:
refactor: 예외처리 로직 단순화 및 패키지 분리

cookie-meringue committed:
feat: 요청한 리소스를 찾을 수 없을 때 발생하는 예외를 처리하는 Resolver 구현

cookie-meringue committed:
refactor: 예외 상속 계층을 고려하도록 변경

- 시간복잡도 O(1)에서 O(d) (d: 예외 상속 depth)로 늘어나지만, d는 일반적으로 그리 크지 않음(평균적으로 5 정도).

cookie-meringue committed:
test: 예상 가능 예외 Resolver 구현에 따른 테스트 코드 수정

cookie-meringue commented:
indent 크기가 잘못 적용되었습니다.
indent 크기는 기존 값으로 수정하겠습니다.

cookie-meringue committed:
style: IDE의 자동 포매팅으로 인해 의도치 않고 변경된 코드 롤백

cookie-meringue committed:
docs: 로컬 테슽 과정에서 잘못 설정한 local 설정 제거

cookie-meringue committed:
refactor: 예상 가능 예외를 warn 으로 로깅하도록 변경

- 이전 리팩터링 과정에서 BusinessExceptionResolver 에서만 예외를 로깅하고 있었음.

threepebbles commented:
변경 사항이 많군요. 코드래빗의 도움을 조금 받아보죠.
@coderabbitai

threepebbles commented:
@coderrabbitai
아래의 내용을 중점으로 리뷰 부탁해

1) 예외 처리 재설계
- `@ExceptionHandler` 방식에서 사용자 구현 `ExceptionResolver` 사용 방식으로 변경
- `@ExceptionHandler` 방식에 비교해 어떤 장단점이 있는지

2) Facade/Composite 패턴 도입
- `ExceptionResolverFacade`에서 Upper Bounded Wildcard를 사용한 방식(`Map<Class<? extends Exception>, ExpectedExceptionResolver<? extends Exception>> expectedExceptionResolvers`)으로 Exception 핸들링

threepebbles commented on backend/spring-routie/src/main/java/routie/global/exception/infrastructure/ExceptionResolverFacade.java:
```diff
@@ -0,0 +1,65 @@
+    public ExceptionResolverFacade(
+            final List<ExpectedExceptionResolver<? extends Exception>> expectedExceptionResolvers,
+            final UnexpectedExceptionResolver unexpectedExceptionResolvers) {
+        this.expectedExceptionResolvers = expectedExceptionResolvers.stream()
+                .collect(Collectors.toMap(
+                        ExceptionResolver::getResolvableException,
+                        Function.identity()
+                ));
+        this.unexpectedExceptionResolvers = unexpectedExceptionResolvers;
```
중복된 예외에 대한 Resolver가 존재하는지 체크해주면 좋을 것 같네요!

threepebbles commented on backend/spring-routie/src/main/java/routie/global/exception/infrastructure/resolver/expected/ExpectedExceptionResolver.java:
```diff
@@ -0,0 +1,43 @@
+
+    private final Class<T> exceptionClass;
+
+    protected ExpectedExceptionResolver() {
+        Type genericSuperclass = getClass().getGenericSuperclass();
+        if (genericSuperclass instanceof final ParameterizedType parameterizedType) {
+            @SuppressWarnings("unchecked")
+            Class<T> exceptionType = (Class<T>) parameterizedType.getActualTypeArguments()[0];
+            this.exceptionClass = exceptionType;
+        } else {
+            throw new BeanInitializationException("예상 가능 예외 핸들러의 제네릭 타입이 잘못 명시되었습니다.");
+        }
+    }
```
스프링 빈이 CGLIB 프록시 객체를 사용하기 때문에 발생하는 문제군요.
기존 코드에서는 ExpectedExceptionResolver 구현체에서 원하는  genericSuperclass를 얻어오지 못할  것 같습니다.
코드래빗의 제안 코드는 다소 복잡해보이네요.

```java
public abstract class ExpectedExceptionResolver<T extends Exception> implements ExceptionResolver {
    private final Class<T> exceptionClass;
    protected ExpectedExceptionResolver(Class<T> exceptionClass) {
        this.exceptionClass = exceptionClass;
    }
}
```

```java
@Component
final class XXXExceptionResolver extends ExpectedExceptionResolver<IllegalArgumentException> {
    XXXExceptionResolver() { super(IllegalArgumentException.class); }
    ...
}
```
리플렉션에 의존하지 않고, 이런 식으로 단순하게 해결하는 방법도 있을 것 같습니다.

cookie-meringue commented:
ㅋㅋㅋㅋㅋㅋ 부끄러운데요

threepebbles reviewed (commented):
수고하셨어요 머랭!

성능(시간 복잡도와 공간 복잡도) 관점에서는 `@ExceptionHandler`와 거의 같군요.
이번 PR의 설계를 가져가면 예외 종류마다 하나의 예외 처리를 위한 클래스가 늘어나서 프로그램 복잡도는 다소 늘어나겠지만, 예외 처리 내부 로직을 세밀하게 커스터마이징할 수 있다는 이점이 있습니다. 다른 팀원들과 함께 유지보수만 잘 할 수 있다면 좋은 시도가 될 것 같네요 :)

남긴 코멘트들만 한번씩 확인해주세요~

cookie-meringue commented:
_⚠️ Potential issue_

**resolve(...) 제네릭 캡처 문제 — 현재 형태로 타입 안전하지 않음**

ExpectedExceptionResolver<? extends Exception>의 resolve(E)에 Exception을 직접 전달할 수 없어 컴파일 타임 타입 불일치가 발생합니다. 안전 캐스팅 헬퍼로 위임하세요.


```diff
-            return findMostSpecificResolver(exception)
-                    .map(expectedExceptionResolver -> expectedExceptionResolver.resolve(exception))
-                    .orElseGet(() -> unexpectedExceptionResolvers.resolve(exception));
+            return findMostSpecificResolver(exception)
+                    .map(resolver -> resolveSafely(resolver, exception))
+                    .orElseGet(() -> unexpectedExceptionResolvers.resolve(exception));
```

헬퍼 메서드 추가(파일 내 임의 위치, private):

```java
private <E extends Exception> ExceptionDetail resolveSafely(
        final ExpectedExceptionResolver<E> resolver,
        final Exception exception
) {
    return resolver.resolve(resolver.getResolvableException().cast(exception));
}
```

<!-- fingerprinting:phantom:medusa:chinchilla -->

<!-- This is an auto-generated reply by CodeRabbit -->

---

코드래빗의 코멘트인데요, 컴파일 타임 타입 불일치라고 하는데, 컴파일에 아무런 문제가 없습니다.
```java
private <E extends Exception> ExceptionDetail resolveSafely(
        final ExpectedExceptionResolver<E> resolver,
        final Exception exception
) {
    return resolver.resolve(resolver.getResolvableException().cast(exception));
}
```

이 메서드가 어떤 문제를 해결하기 위해 존재해야 하는 지 이해가 어려워 질문드립니다~

코드래빗이 `ExpectedExceptionResolver` 생성자 내부에 있는 제네릭 타입 캡쳐를 확인하지 못하고 작성한 리뷰가 아닐까 싶습니다.

threepebbles commented:
> _⚠️ Potential issue_
> 
> **resolve(...) 제네릭 캡처 문제 — 현재 형태로 타입 안전하지 않음**
> 
> ExpectedExceptionResolver<? extends Exception>의 resolve(E)에 Exception을 직접 전달할 수 없어 컴파일 타임 타입 불일치가 발생합니다. 안전 캐스팅 헬퍼로 위임하세요.
> 
> 
> ```diff
> -            return findMostSpecificResolver(exception)
> -                    .map(expectedExceptionResolver -> expectedExceptionResolver.resolve(exception))
> -                    .orElseGet(() -> unexpectedExceptionResolvers.resolve(exception));
> +            return findMostSpecificResolver(exception)
> +                    .map(resolver -> resolveSafely(resolver, exception))
> +                    .orElseGet(() -> unexpectedExceptionResolvers.resolve(exception));
> ```
> 
> 헬퍼 메서드 추가(파일 내 임의 위치, private):
> 
> ```java
> private <E extends Exception> ExceptionDetail resolveSafely(
>         final ExpectedExceptionResolver<E> resolver,
>         final Exception exception
> ) {
>     return resolver.resolve(resolver.getResolvableException().cast(exception));
> }
> ```
> 
> <!-- fingerprinting:phantom:medusa:chinchilla -->
> 
> <!-- This is an auto-generated reply by CodeRabbit -->
> 
> ---
> 
> 코드래빗의 코멘트인데요, 컴파일 타임 타입 불일치라고 하는데, 컴파일에 아무런 문제가 없습니다.
> ```java
> private <E extends Exception> ExceptionDetail resolveSafely(
>         final ExpectedExceptionResolver<E> resolver,
>         final Exception exception
> ) {
>     return resolver.resolve(resolver.getResolvableException().cast(exception));
> }
> ```
> 
> 이 메서드가 어떤 문제를 해결하기 위해 존재해야 하는 지 이해가 어려워 질문드립니다~
> 
> 코드래빗이 `ExpectedExceptionResolver` 생성자 내부에 있는 제네릭 타입 캡쳐를 확인하지 못하고 작성한 리뷰가 아닐까 싶습니다.

저도 컴파일 에러는 발생하지 않았어요. 그 부분은 코드래빗이 헷갈린 것 같네요.
그와 별개로 제가 코멘트 드린 문제는 런타임에 발생할 수 있는 문제에요!

> ExceptionResolver<T extends Exception>에서의 T 타입과, public final ExceptionDetail resolve(final Exception exception)에서 exception의 실제 타입이 다른 타입 불일치 문제가 발생할 수 있군요.

예를 들면 `ExpectedExceptionResolver<AliceException> resolver`가 `resolver.resolve(new BobException());`을 호출할 수 있는 문제가 발생할 수 있어요.
지금은 resolve()를 호출하는 쪽에서 타입 일치를 확인하고 사용하는 방식이 아닌, resolve() 메서드 자체를 안전하게 사용하기 위한 피드백으로 생각했습니다.

goohong reviewed (approved):
👍
> 예외 핸들링 로직이 하나의 클래스(GlobalControllerAdvice) 내에 전부 존재했음. 이로 인해 가독성이 저해되고, SRP 를 준수하지 못하고 있었음.

정교한 예외 처리 구조를 설계하시느라 고생 많으셨습니다. Composite 패턴과 OCP 원칙까지 적용하며 전체적인 설계 의도에 대해 깊이 고민하신 것 같습니다.

리뷰하면서 한 가지 궁금한 점이 생겼습니다. 새로운 구조가 설계적으로 뛰어나지만, 기존의 일반적인 `@ExceptionHandler` 방식에 비해 처음 접하는 개발자에게는 학습 곡선이 가파를 수 있겠다는 생각이 들었습니다.

혹시 일반적인 접근 방식, 기존의 `GlobalExceptionHandler`와 유사하게(`BusinessException` 같은 공통 부모 예외를 두고 `InvalidParamException`, `NotFoundException` 등으로 상속하여 계층을 만든 뒤, `@ControllerAdvice`에서 각 예외 타입에 맞춰 `@ExceptionHandler`로 처리하는 방식)과 비교했을 때, 어떤 특장점이 있는 지 더 풀어서 설명해주실 수 있을까요? 분명 이유가 있을 것 같은데, 코드만 보아서는 그 점을 저는 아직 잘 이해하지 못한 것 같습니다. 혹은 이전에 합의한 사항이 있는 데 기억이 안 나는 걸지도 모르겠네요. 😂

물론 현재 제안주신 구조가 '관심사의 분리(SoC)'나 '일관된 로깅 정책 강제' 같은 확실한 이점을 제공한다는 점은 이해하고 있습니다. 다만, 이 이점들이

- 기존 방식으로는 해결 불가능한 점인가? 와
- 앞서 말씀드린 '학습 비용 증가'라는 단점을 감수할 만큼의 가치가 있다고 판단하신 핵심적인 이유 (+ 바퀴를 다시 만들어야할 이유?)

가 궁금합니다.

cookie-meringue commented:
`resolveSafely` 와 별개로, CGLIB 프록시 적용 시 final 키워드로 인한 문제가 발생했습니다.

## CGLIB 프록시 적용 시 final 키워드로 인한 문제
CGLIB 는 상속 기반 프록시 생성 라이브러리임.
프록시를 적용해야 하는 메서드를 오버라이드해서 동작함
이로 인해, 프록시 사용 시 추상 클래스(`ExpectedExceptionResolver`)의 템플릿 메서드에 fianl로 인해 메서드를 오버라이드 하지 못해 프록시 객체가 null로 초기화되는 문제 발생.

해결 방법 모색 중입니다. 의견 주시면 감사하겠습니다.

cookie-meringue committed:
feat: 예상 가능 예외 Resolver 가 중복으로 등록되지 않도록 보장

- BeanDefinition 을 기반으로, Bean 초기화 시 수행.
- 같은 예외 클래스에 대해 중복으로 등록된 ExpectedExceptionResolver 가 존재한다면, 애플리케이션이 실행되지 않음.

cookie-meringue committed:
refactor: 더 적절한 에외를 던지도록 변경

- BeanInitializationException은 스프링의 기술.
- ExpectedExceptionResolver의 구체 클래스가 Bean 이라는 보장을 할 수 없음.

cookie-meringue committed:
refactor: 올바르지 않은 Exception 전달 시, 예외를 발생시키도록 변경

- 명시적 처리를 위함.

cookie-meringue committed:
fix: 예외 처리 실패 여부가 제대로 로깅되지 않던 문제 수정

cookie-meringue committed:
style: 들여쓰기 수정

cookie-meringue commented:
# 같은 예외에 대해 중복 ExpectedExceptionResolver 가 등록되지 않도록 구현했습니다.

BeanDefinition 을 사용해 Bean 초기화 시 모든 ExpectedExceptionResolver를 탐색하며 중복 ExpectedExceptionResolver가 존재하는지 검사합니다.

만약, IllegalArgumentException 을 다루는 
```java
class AExpectedExceptionResolver  extends ExpectedExceptionResolver<IllegalArgumentException>{
}

class BExpectedExceptionResolver  extends ExpectedExceptionResolver<IllegalArgumentException>{
}
```
가 존재한다면, 애플리케이션이 실행되지 않습니다.

cookie-meringue committed:
chore: 패키지 구조 재정비

cookie-meringue committed:
chore: 패키지명 오타 수정

cookie-meringue committed:
refactor: 예외 처리 실패 시 이유를 더 명확히 로깅하도록 변경

DongchannN reviewed (commented):
어떤 예외를 추가해야겠다 라는 생각이 들었을 때, ExceptionController에 들어가는 것이 일반적인 생각이 들 것 같아요.

머랭의 설계가 훌륭하고, 코드를 알면 유지보수하기 편하다는 장점이 있는데 동의합니다. 하지만 고수준의 설계는 팀원들이 따라가기 힘들다고 생각해요. 코드를 리뷰하는 입장에서 자세히 보아도 작성자처럼 이해하는데 시간도 걸리고 학습도 덜 되었으니까요.

일단 저희의 팀 규모가 적으니 문서화나 대면으로 코드를 설명하는 방향으로 해결해보도록 하죠.

해당 코드가 어느 부분이 위험한지, 사용할 때 어떤 것을 주의하며 사용해야하는 지 머랭의 설명을 듣고 검토한 뒤 Merge하는 것은 어떤가요?

cookie-meringue committed:
fix: 패키지 분리로 인해 로깅이 제대로 동작하지 않던 문제 수정

cookie-meringue committed:
style: 가독성있게 개행하도록 변경

cookie-meringue committed:
chore: 사용하지 않는 `@Slf4j` 어노테이션 제거

cookie-meringue committed:
refactor: dto ui 변환 로직을 Controller 에서 수행하도록 변경

cookie-meringue committed:
refactor: ExceptionResolver 구현 구조를 자연스럽게 변경

- UnexpectedExceptionResolver 가 자신이 어떤 예외를 처리하는지 아는 것이 부자연스러웠음.

cookie-meringue committed:
chore: 계층을 더욱 명확하게 분리

DongchannN reviewed (approved):
Resolver를 도메인에 두고, Service에서 도메인을 의존하게 하셨네요 👍 
기존보다 훨씬 직관적이고 이해하기 쉬운 것 같아요. 고생하셨습니다

cookie-meringue committed:
feat: 예외 처리에 필요한 여러 인지를 주고받을 수 있도록 변경

- 기존엔 Exception 만 주고받을 수 있었음.
- Exception 을 ExceptionContext 내부로 이동.
- 추후 다른 값들을 주고받기에 용이.

cookie-meringue committed:
chore: 클래스 이름에 역할이 더욱 잘 드러나도록 변경

cookie-meringue commented:
@coderabbitai

`ExceptionResolvingRequest` 로 리팩터링한 부분에 대해 리뷰해.
`ExceptionResolvingRequest` 를 분리한 이유는, 예외를 Resolve하는 데에는 `Exception` 말고 더욱 다양한 정보가 필요할 수 있기 때문이야.
예를 들면, HttpServletRequest 의 요청 ip 정보가 있어.

이런 부가적인 정보는 기존의 `ExceptionResolver` 에서는 받을 수 없었어. `Exception`만 매개변수로 전달받도록 시그니처가 구성되어있기 때문이야.

`@RestControllerAdvice`에서 파라메터바인딩으로 얻을 수 있는 모든 정보를 대비해 시그니처를 확장할 수는 없으므로, `ExceptionResolvingRequest` 로 추상화했어.

이제 만약, 예외를 Resolve하는데 더 많은 정보가 필요하다면, `ExceptionResolvingRequest`에 필드를 추가하고 `ExceptionController` 에서 해당 정보를 넘겨 주도록 구현만 하면 돼.

getter 로 각자가 필요한 정보만 선택적으로 꺼냄으로써, 하위호환성을 완벽히 보장할 수 있어. 중간에 `ExceptionResolvingRequest`의 필드가 추가되어도, 기존 로직인 전혀 영향을 받지 않아.


이 설계에 대해 평가해.
그리고, `ExceptionResolvingRequest`를 구현하기 위해 제네릭을 추가적으로 사용했는데, 이 부분에 대해서도 평가해.
문제점이 있다면 말해줘.

cookie-meringue commented:
@threepebbles @goohong @DongchannN 

지금까지의 구현을 토대로, 제 아키텍처는 롤백하는 것이 옳다고 판단했습니다.
"`ExpectedExceptionResolver`의 고정된 시그니처로 인한 유연성 감소" 문제를 해결하기 위해서는 너무나도 복잡한 설계를 적용해야 한다는 것을 깨달았습니다.
제가 작성한 코드가 점점 거대한 프레임워크가 되어 가는 모습을 보니 @goohong 가 우려했던 점들이 발견되었습니다. "이렇게까지 설계를 해야 할까?" 하는 생각이 드네요.
**"머랭이 주말 동안 열심히 작업했으니까"** 같은 이유는 저희 프로젝트에 제 코드를 받아들이는 데에 이유가 될 수 없다고 생각합니다.
따라서, 각 예외 별 핸들링 로직을 제외한 작업 내용 대부분을 폐기하고 다시 설계하고자 합니다.

이번 리팩터링에서 적용했던 재미있는 아이디어와 좋은 설계는 개인 프로젝트를 만들어 프레임워크로 제작해 보겠습니다.

좋은 설계와 클린 코드에 대한 제 욕심이 과했던 점 인정하겠습니다. 제 욕심으로 인해 복잡한 코드와 설계를 파악하게 했던 점 사과드립니다.

---
이제, 복잡한 설계를 가져가지 않고 어떻게 해결할 수 있을 지에 대한 제 아이디어를 제시하겠습니다.

이번 PR에서 제가 해결하고 싶던 문제는 두 가지였습니다.
1. 한 클래스에 여러 `@ExceptionHandler` 들이 몰려 있어 가독성이 떨어지는 문제.
2. `@ExceptionHandler`에서 어노테이션과 파라메터의 예외 타입이 동일하도록 강제할 수 없다는 문제.

첫 번째 문제는 현재 설계가 아니어도 클래스 분리를 통해 해결할 수 있습니다.
문제는 두 번째인데, 이 부분은 커스텀 어노테인션과 어노테이션 프로세서를 통해 해결할 수 있을 것 같습니다.

내일부터 작업 시작하겠습니다.

cookie-meringue committed:
chore: 유연성을 위한 템플릿 인터페이스, Facade 구조 제거

cookie-meringue committed:
feat: ErrorCode가 ProblemDetail을 생산할 수 있도록 구현

cookie-meringue committed:
feat: 런타임 예외를 방지하기 위한 `@ExceptionHandler` 메서드 검증기 구현

- `@ExceptionHandler`를 사용했을 때, 런타임 예외가 발생할 수 있는 상황들을 미연에 방지.
- ApplicationContext가 초기화될 때 검증 수행.

cookie-meringue committed:
feat: 각 예외 별 핸들러 추가

cookie-meringue committed:
feat: 예외 핸들러에서 예외 발생 시 Fallback 구현

cookie-meringue committed:
refactor: 예외처리 방식 변경에 따른 pointcut 변경

cookie-meringue committed:
Merge branch 'develop' into refactor/structuring-exception-handler-#723
# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

cookie-meringue committed:
test: 예외 처리 추가에 따른 테스트 코드 변경

cookie-meringue committed:
fix: 잘못된 접근 제어자 롤백

cookie-meringue commented:
@DongchannN @goohong @threepebbles 

새로 구현한 아키텍처를 정리하고 문서화했습니다. 확인 부탁드립니다!
노션(backend - 기술 기록)에도 작성했으니 편한 방식으로 확인 부탁드립니다.

# 기존 예외 처리 설계의 문제점

## 기존 예외 처리 설계

하나의 `@RestControllerAdvice`클래스에 모든 `@ExceptionHandler`들이 모여 있었음.

```java
@RestControllerAdvice
class GlobalExceptionHandler {
|
|--- @ExceptionHandler(BusinessException.class)
|
|--- @ExceptionHandler(HandlerMethodValidationExcpetion.class)
|
|--- //...
|
|--- @ExceptionHandler(Exception.class)
}
```

## 문제점 1

특정 상황에 쓰이는 `@ExceptionHandler`를 찾는 과정이 불편함.

- 특정 상황(HandlerMethod의 파라메터 관련 문제 상황)을 지원하는 `@ExceptionHandler` 탐색 과정 예시.
    1. Command + F로 “HandlerMethod” 를 입력 후 HandlerMethod 관련 `@ExceptionHandler`들이 모여 있는 코드로 이동.
    2. `MethodArgumentTypeMismatchException`을 지원하는 `@ExceptionHandler` 발견.
    3. 해당 `@ExceptionHandler`의 javadoc을 읽은 후, 원하는 `@ExceptionHandler`가 아니었음을 인지.
    4. 스크롤을 내려 다른 `@ExceptionHandler`들을 찾기 시작.
    5. 지금까지 어떤 `@ExceptionHandler`들을 넘겼는지 잊어버림.
    6. 결국, 다시 위로 돌아가 다시 하나씩 탐색 시작.
    7. javadoc으로 인해 코드 라인 수가 많기 때문에, 피로감을 느낌.

## 문제점 2

새로운 `@ExceptionHandler`를 추가하는 과정이 불편함.

- 새로운 `@ExceptionHandler`(`IllegalArgumentException`)를 추가하는 과정 예시.
    1. `GlobalExceptionHandler` 내부에 `IllegalArgumentException`을 지원하는 `@ExceptionHandler`가 있는지 확인.(있다면 중복이므로)
    2. 모든 `@ExceptionHandler`에는 javadoc 주석이 작성되어있을 것이므로, 내려야 하는 스크롤의 양이 많음.
    3. `IllegalArgumentException`을 지원하는 `@ExceptionHandler`가 없다면, 적절해보이는 위치에 새로운 `@ExceptionHandler`를 추가해야 함.
    4. 어떤 위치에 `@ExceptionHandler`를 추가해야 할 지 판단하기 어렵고, 개개인의 가치관마다 새로운 `@ExceptionHandler`의 삽입 위치가 달라짐.
    5. `@ExceptionHandler`의 순서 일관성이 없어져 탐색에 어려움을 겪게 됨.

## 문제점 3

`@ExceptionHandler` 내부의 예외 클래스와 메서드 파라메터의 클래스 타입을 일치시킬 수 없음.

```java
@ExceptionHandler(IllegalArgumentException.class)
public ProblemDetail handle(NullPointerException exception) {
	// handle
}
```

위 코드 동작 시, 어노테이션과 파라메터의 예외 타입이 맞지 않아 런타임에 예외가 발생함.(예외 핸들러에서 예외 발생)

AOP 등을 사용해 예외 핸들러에서 발생한 예외를 핸들링할 수는 있지만, 이런 상황이 발생하는 것은 운영 환경이 아닌 테스트 과정에서 전부 끝나야 한다고 생각함.

어노테이션과 파라메터의 예외 타입이 동일하도록 강제할 수단이 존재하지 않음.

## 문제점 4

한 예외에 대해 여러 `@ExceptionHandler`가 존재할 수 있음.

Spring MVC 는 패키지 레벨에 따른 예외 핸들링을 구현할 수 있게 하기 위해 이를 허용함.

그러나, 우리 프로젝트에서는 중앙 집중화된 방식으로 예외를 처리하도록 설계했음.

→ 패키지 레벨에 따른 예외 핸들링이 불필요!

이러한 상황에서, 한 예외에 대해 여러 `@ExceptionHandler`가 존재할 수 있다는 점은 오히려 방해 요소가 됨.

예외를 처리하는 패키지가 한 군데이기 때문.

한 예외에 대해 여러 `@ExceptionHandler`를 만들면 예외 처리 동작을 파악하기 매우 힘듦.

# 해결 방안

## ❌ 해결 방안 1 - 예외 처리기 추상화를 통한 예외 처리 핸들러의 템플릿화

예상 가능한 예외와 예상 불가능한 예외를 추상화.

예상 가능한 새로운 예외 추가 시 `예상 가능 예외 처리기` 를 상속받는 새로운 `Resolver` 를 구현하고 Bean 으로 등록하면, 자동으로 추가되도록 구현. (Spring 의 List DI 를 활용)

예상 불가 예외 처리기는 `Exception`을 받아 로깅하도록 구현.

`예외 처리기 Facade`를 통해 예상 가능한 예외와 예상 불가능한 예외를 적절히 핸들링하도록 구현.

- `DispatcherServlet` 과 같은 원리로 동작.
- 하나의 `@ExceptionHandler`가 최상단에서 예외를 캐치.
- 캐치한 예외를 `예외 처리기 Facade`로 처리 위임.
    - 예외 처리기 Facade 내부 동작.
    1. 예상 가능 예외로 등록된 예외인 경우, 해당 예외 처리기를 통해 처리
    2. 예상 가능 예외로 등록되지 않은 예외인 경우, 예상 불가 예외 처리기를 통해 처리

<img width="6196" height="3600" alt="image" src="https://github.com/user-attachments/assets/c808a54a-9620-4223-acdd-a4e8aa5badf1" />

### 해결 방안 1의 장점

제네릭 타입 캡쳐를 통해 `@ExceptionHandler` 내부의 예외 클래스와 메서드 파라메터의 클래스 타입 일치를 “컴파일 단계”에 강제 가능.

```java
abstract class ExpectedExceptionResolver<T extends Exception> {

    private final Class<T> exceptionClass;

    protected ExpectedExceptionResolver() {
	    // T 타입 캡쳐를 통해 exceptionClass를 초기화.
    }

    public ExceptionDetail resolve(final Exception exception) {
	    // 생략.
      return esolveInternal(exception);
    }

    protected abstract ExceptionDetail resolveInternal(T exception);
}

@Component
class SomeExceptionResolver extends ExpectedExceptionResolver<SomeException> {

		// 컴파일 에러 발생!!
		// 제네릭 클래스와 매개변수 클래스가 일치하지 않기 때문.
    @Override
    protected ExceptionDetail resolveInternal(AnotherException exception) {
        return ExceptionDetail.fromErrorCode(INVALID_REQUEST_DATA_VALUE);
    }
}
```

개발자가 새로운 `ExceptionResolver`를 추가할 때 실수할 일이 없음.

### 해결 방안 1의 한계

예상 가능 예외 처리기로 추상화하기 위해서는, 같은 메서드 시그니처를 자식 클래스들이 공유해야 함.

자식 클래스들은 `abstract ProblemDetail resolveInternal(Exception exception);` 을 구현해야 함.

그러나, Spring 은 `@ExceptionHandler` 에 여러 가지 파라메터를 바인딩하는 기능을 제공함.

`@ExceptionHandler` 메서드는 `Exception` 뿐만 아니라, `HttpServletResponse` 등을 파라메터로 바인딩받을 수 있음.

해결 방안 1은, `Exception`만을 매개변수로 받아 핸들링할 수 있음.

이 문제점을 해결하기 위해 예외 처리를 위한 Context 를 담는 Container 레코드를 고려.

```java
record ExceptionContext<T extends Exception>(
        T exception,
        HttpServletRequest request,
        HttpServletResponse response,
        // ...
) {
}
```

구현에 성공했으나, 설계가 지나치게 복잡해지는 문제 발견.

 ⇒ 해결 방안 1의 구조로는 이를 해결할 수 없다고 판단.

## ⭕ 해결 방안 2 - 각 예외 핸들러를 클래스로 분리

하나의 `@RestControllerAdvice` 클래스 아래에 있던 `@ExceptionHandler`들을 각각의 `@RestControllerAdvice - @ExceptionHandler` 쌍으로 분리.

**기존 구조**

```java
@RestControllerAdvice
class GlobalExceptionHandler {
|
|--- @ExceptionHandler(BusinessException.class)
|
|--- @ExceptionHandler(HandlerMethodValidationExcpetion.class)
|
|--- //...
|
|--- @ExceptionHandler(Exception.class)
}
```

**변경 후 구조**

```java
@RestControllerAdvice
class BusinessExceptionHandler {
|
|--- @ExceptionHandler(BusinessException.class)
}

@RestControllerAdvice
class HandlerMethodValidationExcpetionHandler {
|
|--- @ExceptionHandler(HandlerMethodValidationExcpetion.class)
}

@RestControllerAdvice
class UnexpectedExcpetionHandler {
|
|--- @ExceptionHandler(Exception.class)
}
```

### 해결 방안 2의 장점

시그니처의 제약이 없음.

```java
@RestControllerAdvice
public class AExceptionHandler {

		// 매개변수로 아무것도 받지 않을 수 있음.
    @ExceptionHandler(AException.class)
    public ProblemDetail handle() {
        return ErrorCode.A.toProblemDetail();
    }
}

@RestControllerAdvice
public class BExceptionHandler {

		// 매개변수로 Exception을 받을 수 있음.
    @ExceptionHandler(BException.class)
    public ProblemDetail handle(BException exception) {
        ExceptionLogger.logExpectedException(exception);
        return exception.getErrorCode().toProblemDetail();
    }
}

@RestControllerAdvice
public class CExceptionHandler {

		// 매개변수로 Exception, HttpServletRequest를 받을 수 있음.
    @ExceptionHandler(CException.class)
    public ProblemDetail handle(CException exception, HttpServletRequest req) {
        ExceptionLogger.logExpectedException(exception);
        // req를 다루는 로직.
        return exception.getErrorCode().toProblemDetail();
    }
}
```

상속과 구현의 관계에서 벗어났기 때문에, 예외 핸들러 시그니처가 다를 수 있음.

### 해결 방안 2의 한계

사실상 클래스 분리만 한 것이므로, 기존의 문제점을 전부 해결할 수 없음.

**해결할 수 없는 문제 1**

> `@ExceptionHandler` 내부의 예외 클래스와 메서드 파라메터의 클래스 타입을 일치시킬 수 없음.
> 
> 
> ```java
> @ExceptionHandler(IllegalArgumentException.class)
> public ProblemDetail handle(NullPointerException exception) {
> 	// handle
> }
> ```
> 
> 위 코드 동작 시, 어노테이션과 파라메터의 예외 타입이 맞지 않아 런타임에 예외가 발생함.(예외 핸들러에서 예외 발생)
> 
> AOP 등을 사용해 예외 핸들러에서 발생한 예외를 핸들링할 수는 있지만, 이런 상황이 발생하는 것은 운영 환경이 아닌 테스트 과정에서 전부 끝나야 한다고 생각함.
> 
> 어노테이션과 파라메터의 예외 타입이 동일하도록 강제할 수단이 존재하지 않음.
> 

**해결할 수 없는 문제 2**

> 한 예외에 대해 여러 `@ExceptionHandler`가 존재할 수 있음.
> 
> 
> Spring MVC 는 패키지 레벨에 따른 예외 핸들링을 구현할 수 있게 하기 위해 이를 허용함.
> 
> 그러나, 우리 프로젝트에서는 중앙 집중화된 방식으로 예외를 처리하도록 설계했음.
> 
> → 패키지 레벨에 따른 예외 핸들링이 불필요!
> 
> 이러한 상황에서, 한 예외에 대해 여러 `@ExceptionHandler`가 존재할 수 있다는 점은 오히려 방해 요소가 됨.
> 
> 예외를 처리하는 패키지가 한 군데이기 때문.
> 
> 한 예외에 대해 여러 `@ExceptionHandler`를 만들면 예외 처리 동작을 파악하기 매우 힘듦.
> 

### 해결 방안 2의 한계 해결

스프링 애플리케이션은 세 가지 시점이 있음.

1. 컴파일 시점
2. `ApplicationContext` 초기화 시점 (런타임 시점이나, 실질적 런타임 이전)
3. 실질적 런타임 시점

컴파일 시점에 검증을 수행할 수 없다면, `ApplicationContext` 초기화 시점에 검증하도록 구현.

실질적 런타임에 예외가 발생하는 것 보다는, `ApplicationContext` 초기화 실패로 애플리케이션이 구동되지 않도록 방지하는 구현을 선택.

Spring 의 `BeanPostProcessor`를 통해 빈 초기화 시 검증이 수행되도록 구현.

`@RestControllerAdvice`가 붙은 클래스의 메서드에 대해 다음과 같은 검증 프로세스 수행.

> 1. `@ExceptionHandler` 의 예외 클래스와 메서드의 파라메터에 있는 예외 클래스는 반드시 같아야 한다.
> 
> - 메서드가 예외 클래스를 파라메터로 받지 않는 것은 허용한다.
> - 만약, 메서드가 예외 클래스를 파라메터로 받는다면, 반드시 `@ExceptionHandler` 의 예외 클래스만을 받아야 한다.
> 
> 2. 애플리케이션 전역에서 특정 예외에 대한 `@ExceptionHandler` 는 단 하나만 존재해야 한다.
> 
> - `IllegalArgumentException`을 다루는 `@ExceptionHandler` 는 애플리케이션 전역에서 단 하나만 존재해야 한다.
> - 상속은 고려하지 않는다. (`IllegalArgumentException`을 다루는 핸들러와 `Exception`을 다루는 핸들러는 동시에 존재할 수 있다)

`BeanPostProcessor`를 통한 `ApplicationContext` 초기화 시점 검증을 통해 문제를 해결.

# 새로운 예외 핸들러를 추가하는 방법

`routie.global.exception.infrastructure.expected` 아래에 예외 클래스를 추가.

예외 클래스에는 반드시 `@RestControllerAdvice` 어노테이션을 사용.

혼동을 방지하기 위해 하나의 예외 클래스에는 하나의 `@ExceptionHandler`만 존재하도록 구현.

`IllegalArgumentException` 핸들러 추가 예시

```java
/**
* 설명
*/
@RestControllerAdvice
public class IllegalArgumentExceptionHandler {

    @ExceptionHandler(IllegalArgumentException.class)
    public ProblemDetail handle(final IllegalArgumentException exception) {
        ExceptionLogger.logExpectedException(exception);
				return ErrorCode.ILLEGAL_ARGUMENT.toProblemDetail();
    }
}
```

예외 클래스에 반드시 javadoc 을 통해 어떤 예외를 핸들링하는지, 예시 상황은 무엇이 있는지 작성해야 함.

# 기대 효과

## 1. 가독성

`@ExceptionHandler` 를 디렉터리 구조로 관리할 수 있어 가독성과 명시성이 매우 높아짐.

<img width="456" height="433" alt="image" src="https://github.com/user-attachments/assets/5a6e534a-369f-4762-89e2-b02103b5a126" />

각 예외 타입별로 패키지 분리 가능.

하나의 클래스에서 특정 예외 핸들러를 찾는 것 보다, 패키지와 클래스명을 통해 찾을 수 있어 더욱 편리함.

## 2. 런타임 예외 방지

하나의 예외에 대한 `@ExceptionHandler`가 겹침.

`@ExceptionHandler`의 예외 타입과 파라메터의 예외 타입이 다름.

`@ExceptionHandler` 메서드의 파라메터로 여러 `Exception`을 받음.

위 세 가지 경우, 스프링 애플리케이션이 실행될 때 아무런 문제가 없음.

그러나, 실질적으로 해당 `@ExceptionHandler`가 호출될 때 런타임 예외가 발생.

예외를 핸들링하는 핸들러에서 예외가 발생하는 것, 그리고 이를 런타임 전에 발견하지 못한 것은 치명적인 문제라고 판단.

`BeanPostProcessor` 를 통해 해당 상황들을 미연에 방지할 수 있음.

⇒ 런타임 예외 발생 전 사전에 판단할 수 있음(테스트가 실패할 것이기 때문)

cookie-meringue commented:
@coderabbitai
리팩터링한 구조 분석하고 평가한 후, 취약점이 있다면 원인과 예시 형식으로 나열해.

cookie-meringue commented:
코드래빗의 리뷰 내용이 적절하지 못하다 판단해 삭제했습니다.

DongchannN commented on backend/spring-routie/src/main/java/routie/global/exception/infrastructure/config/ExceptionHandlerPostProcessor.java:
```diff
@@ -0,0 +1,120 @@
+                .filter(Throwable.class::isAssignableFrom)
+                .toList();
+
+        if (throwableParameters.size() > 1) {
+            throw new BeanDefinitionValidationException(
+                    "ExceptionHandler 메서드는 최대 하나의 예외 파라미터만 가질 수 있습니다: " +
+                            formatExceptionHandlerName(exceptionHandlerMethod)
+            );
+        }
+
+        if (!throwableParameters.isEmpty() && !exceptionClass.equals(throwableParameters.getFirst())) {
+            throw new BeanDefinitionValidationException(
+                    "ExceptionHandler 메서드의 예외 파라미터는 지정된 예외 클래스와 일치해야 합니다: " +
+                            formatExceptionHandlerName(exceptionHandlerMethod)
+            );
+        }
```
해당 부분의 검증은 `@ExceptionHandler`의 처리 가능한 예외를 하나로 강제하고, 반드시 파라미터와 일치해야한다는 규칙이네요

개발을 하다 보면 다른 예외에 대해서 같은 방식으로 처리하고 싶은 상황이 발생할 수 있을 것 같아요. 정확한 1대1 매핑으로 제약의 강도를 높이는 것 보다 `ExceptionHandler`가 복수의 예외를 다룰 수 있게 허용하고, 상속관계 등을 검증해서 런타임 예외를 해결하는 것에 대해 어떻게 생각하시나요?

DongchannN commented on backend/spring-routie/src/main/java/routie/global/exception/infrastructure/config/ExceptionHandlerPostProcessor.java:
```diff
@@ -0,0 +1,120 @@
+        }
+    }
+
+    private void validateUniqueness(final Method exceptionHandlerMethod) {
+        ExceptionHandler annotation = exceptionHandlerMethod.getAnnotation(ExceptionHandler.class);
+        for (final Class<? extends Throwable> exceptionClass : annotation.value()) {
+            if (exceptionHandlers.containsKey(exceptionClass)) {
+                Method existingMethod = exceptionHandlers.get(exceptionClass);
+                throw new BeanDefinitionValidationException(
+                        exceptionClass.getSimpleName() + " 을 처리하는 ExceptionHandler가 중복되었습니다: " +
+                                formatExceptionHandlerName(existingMethod) + ", " + formatExceptionHandlerName(
+                                exceptionHandlerMethod)
+                );
+            }
+            exceptionHandlers.put(exceptionClass, exceptionHandlerMethod);
+        }
+    }
```
**서비스 예외 현황**
- 도메인이 많지 않아서 복잡한 분리가 불필요
- BusinessException + 메시지 차별화로 비즈니스 예외는 통합 처리
- 같은 예외 타입이 다르게 처리될 가능성이 낮음

위와 같은 상황에서는 복수 핸들러가 실수일 가능성이 높아서 이런 검증이 유의미하다고 생각해요.

하지만 도메인이 많아지고 복잡해져 아래와 같이 같은 예외에 대해 다른 방식으로 처리하고 싶은 경우가 있을 것이라 봅니다.

```java
@RestControllerAdvice("com.example.routie")
public class RoutieExceptionHandler {
    @ExceptionHandler(IllegalArgumentException.class)
    public ProblemDetail handle(IllegalArgumentException e) {
        // 루티 관련 로깅, 메트릭 수집
        return ProblemDetail.of();
    }
}

@RestControllerAdvice("com.example.user")
public class UserExceptionHandler {
    @ExceptionHandler(IllegalArgumentException.class) 
    public ProblemDetail handle(IllegalArgumentException e) {
        // 사용자 관련 로깅, 다른 메트릭
        return ProblemDetail.of();
    }
}
```

하나의 클래스 내부에 같은 예외를 처리하는 핸들러가 있다면 스프링이 예외를 터뜨려주니, 클래스가 다른 경우 basePackages가 겹치지 않는 경우에만 같은 예외 타입의 복수 핸들러를 허용하는 방식은 어떨까요?"

DongchannN reviewed (commented):
👍
